primitives/runtime/src/offchain/rpc.rs

```diff
@@ -0,0 +1,431 @@
+	/// Deadline is used instead of Timeout
+	/// Err is returned in case there is a Http
+	/// or deserializing error.
+	pub fn send_with_deadline(&self, deadline: Timestamp, request: &'a Request) -> RpcResult {
+		// Construct http POST body
+		let request_body = json!({
+			"jsonrpc": request.jsonrpc,
+			"id": request.id,
+			"method": request.method,
+			"params": request.params
+		});
+		let mut body: Vec<&[u8]> = Vec::new();
+		let request_body_slice: &[u8] = &(serde_json::to_vec(&request_body).unwrap())[..];
+		body.push(request_body_slice);
+
+		// Send http POST request
+		let post_request = http::Request::post(self.url, body);
+
+		let pending = match post_request
+			.add_header("Content-Type", "application/json;charset=utf-8")
+			.deadline(deadline)
+			.send().map_err(|_| http::Error::IoError) {
+				Ok(result) => result,
+				Err(e) => return Err(Error::Http(e)),
+		};
+
+		// Hanlde http response
+		let response = match pending.try_wait(deadline)
+			.map_err(|_| http::Error::DeadlineReached) {
+				Ok(result) =>
+					match result {
+						Ok(result_bis) => result_bis,
+						Err(e) => return Err(Error::Http(e)),
+					},
+				Err(e) => return Err(Error::Http(e)),
+		};
+
+		// Return http error if response status is not 200
+		if response.code != 200 {
+			log::warn!("Unexpected status code: {}", response.code);
+			return Err(Error::Http(http::Error::Unknown));
+		}
+
+		// Deserialize response body
+		let response_body_bytes = response.body().collect::<Vec<u8>>();
+		let rpc_result = Self::deserialize_response(&response_body_bytes)?;
+
+		Ok(rpc_result)
+	}
+
+	/// Deserialize a RPC response
+	pub fn deserialize_response(response_body_bytes: &Vec<u8>) -> RpcResult {
+		let response_deserialized: Value = serde_json::from_slice(response_body_bytes)
+			.map_err(|_| Error::Deserializing)?;
+
+		response_deserialized.as_object().ok_or(Error::Deserializing)?;
+
+		let jsonrpc_value = response_deserialized.get("jsonrpc").ok_or(Error::Deserializing)?;
+		let jsonrpc = jsonrpc_value.as_str().ok_or(Error::Deserializing)?;
+
+		let id_value = response_deserialized.get("id").ok_or(Error::Deserializing)?;
+		let id = id_value.as_u64().ok_or(Error::Deserializing)?;
+
+		let result_value = response_deserialized.get("result");
+		let error_value = response_deserialized.get("error");
+
+		if error_value == None && result_value == None {
+			return Err(Error::Deserializing);
+		}
+
+		let error: Option<RpcError> = match error_value {
+			Some(value) => {
+				let code_value = value.get("code").ok_or(Error::Deserializing)?;
+				let code = code_value.as_i64().ok_or(Error::Deserializing)?;
+				let message_value = value.get("message").ok_or(Error::Deserializing)?;
+				let message = message_value.as_str().ok_or(Error::Deserializing)?;
+				let data_value = value.get("data");
+
+				let rpc_error = RpcError {
+					code: code as i32,
+					message: message.as_bytes().to_vec(),
+					data: data_value.cloned()
```

Comment:
We should avoid cloning here and it can be done by using `take` instead of `get` when reading etnries from `serde_json::Value`.

Reply:
`cloned()` removed after using `Value::take()` instead of `Value::get()`